### PR TITLE
fix: catch collectible access non-collectible exception

### LINKED ISSUE #11251: [Issue]: .net7 breaking change collectible-assemblies
### Please describe your bug

A breaking change in .net7 affects `Emby.Server.Implementations.Serialization.MyXmlSerializer`

https://learn.microsoft.com/en-us/dotnet/core/compatibility/core-libraries/7.0/collectible-assemblies

### Reproduction Steps

update plugin build from .net6 (10.8) to .net8 (10.9)
 
related: https://github.com/ConfusedPolarBear/intro-skipper/issues/174

### Jellyfin Version

Unstable (master branch)

### if other:

_No response_

### Environment

```markdown
Docker
```

### Jellyfin logs

```shell
[2024-03-28 23:48:18.715 +01:00] [ERR] [21] Emby.Server.Implementations.ScheduledTasks.TaskManager: Error executing Scheduled Task
System.AggregateException: One or more errors occurred. (A non-collectible assembly may not reference a collectible assembly.)
 ---> System.NotSupportedException: A non-collectible assembly may not reference a collectible assembly.
   at System.Reflection.Emit.RuntimeModuleBuilder.<GetTypeRef>g____PInvoke|13_0(QCallModule __module_native, UInt16* __strFullName_native, QCallModule __refedModule_native, Int32 __tkResolution_native)
   at System.Reflection.Emit.RuntimeModuleBuilder.GetTypeTokenWorkerNoLock(Type type, Boolean getGenericDefinition)
   at System.Reflection.Emit.RuntimeModuleBuilder.GetTypeTokenInternal(Type type, Boolean getGenericDefinition)
   at System.Reflection.Emit.SignatureHelper.AddOneArgTypeHelperWorker(Type clsArgument, Boolean lastWasGenericInst)
   at System.Reflection.Emit.SignatureHelper.AddOneArgTypeHelperWorker(Type clsArgument, Boolean lastWasGenericInst)
   at System.Reflection.Emit.RuntimeILGenerator.DeclareLocal(Type localType, Boolean pinned)
   at System.Xml.Serialization.CodeGenerator.DeclareLocal(Type type, String name)
   at System.Xml.Serialization.CodeGenerator.DeclareOrGetLocal(Type type, String name)
   at System.Xml.Serialization.ReflectionAwareILGen.WriteArrayLocalDecl(String typeName, String variableName, SourceInfo initValue, TypeDesc arrayTypeDesc)
   at System.Xml.Serialization.XmlSerializationWriterILGen.WriteElement(SourceInfo source, ElementAccessor element, String arrayName, Boolean writeAccessor)
   at System.Xml.Serialization.XmlSerializationWriterILGen.WriteElements(SourceInfo source, String enumSource, ElementAccessor[] elements, TextAccessor text, ChoiceIdentifierAccessor choice, String arrayName, Boolean writeAccessors, Boolean isNullable)
   at System.Xml.Serialization.XmlSerializationWriterILGen.WriteMember(SourceInfo source, String choiceSource, ElementAccessor[] elements, TextAccessor text, ChoiceIdentifierAccessor choice, TypeDesc memberTypeDesc, Boolean writeAccessors)
   at System.Xml.Serialization.XmlSerializationWriterILGen.GenerateTypeElement(XmlTypeMapping xmlTypeMapping)
   at System.Xml.Serialization.TempAssembly.GenerateRefEmitAssembly(XmlMapping[] xmlMappings, Type[] types)
   at System.Xml.Serialization.TempAssembly..ctor(XmlMapping[] xmlMappings, Type[] types, String defaultNamespace, String location)
   at System.Xml.Serialization.XmlSerializer.GenerateTempAssembly(XmlMapping xmlMapping, Type type, String defaultNamespace, String location)
   at System.Xml.Serialization.XmlSerializer..ctor(Type type, String defaultNamespace)
   at Emby.Server.Implementations.Serialization.MyXmlSerializer.<>c.<GetSerializer>b__1_0(String _, Type t)
   at System.Collections.Concurrent.ConcurrentDictionary`2.GetOrAdd[TArg](TKey key, Func`3 valueFactory, TArg factoryArgument)
   at Emby.Server.Implementations.Serialization.MyXmlSerializer.GetSerializer(Type type)
   at Emby.Server.Implementations.Serialization.MyXmlSerializer.SerializeToWriter(Object obj, XmlWriter writer)
   at Emby.Server.Implementations.Serialization.MyXmlSerializer.SerializeToStream(Object obj, Stream stream)
   at Emby.Server.Implementations.Serialization.MyXmlSerializer.SerializeToFile(Object obj, String file)
   at ConfusedPolarBear.Plugin.IntroSkipper.Plugin.SaveTimestamps()
   at ConfusedPolarBear.Plugin.IntroSkipper.Plugin.UpdateTimestamps(Dictionary`2 newTimestamps, AnalysisMode mode)
   at ConfusedPolarBear.Plugin.IntroSkipper.ChapterAnalyzer.AnalyzeMediaFiles(ReadOnlyCollection`1 analysisQueue, AnalysisMode mode, CancellationToken cancellationToken)
   at ConfusedPolarBear.Plugin.IntroSkipper.BaseItemAnalyzerTask.AnalyzeItems(ReadOnlyCollection`1 items, CancellationToken cancellationToken)
   at ConfusedPolarBear.Plugin.IntroSkipper.BaseItemAnalyzerTask.<>c__DisplayClass5_0.<AnalyzeItems>b__0(KeyValuePair`2 season)
   at System.Threading.Tasks.Parallel.<>c__DisplayClass43_0`2.<PartitionerForEachWorker>b__1(IEnumerator& partitionState, Int64 timeout, Boolean& replicationDelegateYieldedBeforeCompletion)
--- End of stack trace from previous location ---
   at System.Threading.Tasks.Parallel.<>c__DisplayClass43_0`2.<PartitionerForEachWorker>b__1(IEnumerator& partitionState, Int64 timeout, Boolean& replicationDelegateYieldedBeforeCompletion)
   at System.Threading.Tasks.TaskReplicator.Replica.Execute()
   --- End of inner exception stack trace ---
   at System.Threading.Tasks.TaskReplicator.Run[TState](ReplicatableUserAction`1 action, ParallelOptions options, Boolean stopOnFirstFailure)
   at System.Threading.Tasks.Parallel.PartitionerForEachWorker[TSource,TLocal](Partitioner`1 source, ParallelOptions parallelOptions, Action`1 simpleBody, Action`2 bodyWithState, Action`3 bodyWithStateAndIndex, Func`4 bodyWithStateAndLocal, Func`5 bodyWithEverything, Func`1 localInit, Action`1 localFinally)
--- End of stack trace from previous location ---
   at System.Threading.Tasks.Parallel.PartitionerForEachWorker[TSource,TLocal](Partitioner`1 source, ParallelOptions parallelOptions, Action`1 simpleBody, Action`2 bodyWithState, Action`3 bodyWithStateAndIndex, Func`4 bodyWithStateAndLocal, Func`5 bodyWithEverything, Func`1 localInit, Action`1 localFinally)
   at System.Threading.Tasks.Parallel.ForEachWorker[TSource,TLocal](IEnumerable`1 source, ParallelOptions parallelOptions, Action`1 body, Action`2 bodyWithState, Action`3 bodyWithStateAndIndex, Func`4 bodyWithStateAndLocal, Func`5 bodyWithEverything, Func`1 localInit, Action`1 localFinally)
   at System.Threading.Tasks.Parallel.ForEach[TSource](IEnumerable`1 source, ParallelOptions parallelOptions, Action`1 body)
   at ConfusedPolarBear.Plugin.IntroSkipper.BaseItemAnalyzerTask.AnalyzeItems(IProgress`1 progress, CancellationToken cancellationToken)
   at ConfusedPolarBear.Plugin.IntroSkipper.DetectCreditsTask.ExecuteAsync(IProgress`1 progress, CancellationToken cancellationToken)
   at Emby.Server.Implementations.ScheduledTasks.ScheduledTaskWorker.ExecuteInternal(TaskOptions options)
```

### FFmpeg logs

_No response_

### Please attach any browser or client logs here

_No response_

### Please attach any screenshots here

_No response_

### Code of Conduct

- [X] I agree to follow this project's Code of Conduct

### PR DESCRIPTION
**Pull request by gnattu**
**Changes**

Catch the `NotSupportedException` caused by plugins directly using the server's XML serializer and throw a new one with informative messages, notifying the plugin developer that directly using this serializer is not a supported use case.

This is an alternative to #11254 to retain the in-process restart functionality.

This change could have a long-term impact, meaning that our plugins need to use their own XML serializers from now on, and this change may not be limited to 10.9, but also in later versions.

**Issues**

Fixes #11251

### CODE REVIEW

**Review by Shadowghost — APPROVED** [review]

File: Emby.Server.Implementations/Serialization/MyXmlSerializer.cs [hunk 1]
```diff
@@ -24,6 +24,18 @@ private static XmlSerializer GetSerializer(Type type)
                 static (_, t) => new XmlSerializer(t),
                 type);
 
+        private static void ThrowCollectibleException(Exception e)
+        {
+            var isCollectibleException = string.Equals(e.Message, "A non-collectible assembly may not reference a collectible assembly.", StringComparison.Ordinal);
+            if (isCollectibleException)
+            {
+                throw new NotSupportedException($"It seems you are using the server's XML serializer in your plugin.\n" +
```
Comment (mark-monteiro) [inline Emby.Server.Implementations/Serialization/MyXmlSerializer.cs:32]: You should provide the inner exception as a constructor argument here

File: Emby.Server.Implementations/Serialization/MyXmlSerializer.cs [hunk 2]
```diff
@@ -24,6 +24,18 @@ private static XmlSerializer GetSerializer(Type type)
                 static (_, t) => new XmlSerializer(t),
                 type);
 
+        private static void ThrowCollectibleException(Exception e)
```
Comment (mark-monteiro) [inline Emby.Server.Implementations/Serialization/MyXmlSerializer.cs:27]:
```suggestion
        private static void ThrowCollectibleException(NotSupportedException e)
```

File: Emby.Server.Implementations/Serialization/MyXmlSerializer.cs [hunk 3]
```diff
@@ -24,6 +24,18 @@ private static XmlSerializer GetSerializer(Type type)
...
+        {
+            var isCollectibleException = string.Equals(e.Message, "A non-collectible assembly may not reference a collectible assembly.", StringComparison.Ordinal);
+            if (isCollectibleException)
+            {
+                throw new NotSupportedException($"It seems you are using the server's XML serializer in your plugin.\n" +
+                                                $"This serializer is only used by the server to handle the plugin's config file " +
+                                                $"and is not intended for the plugin to use directly.\n" +
+                                                $"Please create your own XML serializer instance to handle plugin specific files.");
```
Comment (mark-monteiro) [inline Emby.Server.Implementations/Serialization/MyXmlSerializer.cs:35]:
```suggestion
                                                $"Please create your own XML serializer instance to handle plugin specific files.",
                                                e);
```

**Review by crobibero — APPROVED** [review]

**Review by jumoog — APPROVED** [review]

**Comment by mark-monteiro** [comment]
Instead of this, it might be possible to provide an alternate ServiceProvider when constructing the IPlugin instances in the plugin manager. A custom serviceprovider that wraps the standard one, but throws this exception if an attempt is made to resolve the xml serializer. If that works I think it would be a cleaner solution.

**Comment by gnattu** [comment]
> Instead of this, it might be possible to provide an alternate ServiceProvider when constructing the IPlugin instances in the plugin manager. A custom serviceprovider that wraps the standard one, but throws this exception if an attempt is made to resolve the xml serializer. If that works I think it would be a cleaner solution.

How would that work? The `MyXmlSerializer` instance is directly passed into the plugin's constructor

**Comment by cvium** [comment]
Do we have a minimal repro plugin?

**Comment by gnattu** [comment]
> Do we have a minimal repro plugin?

Not exactly minimal. I tried to create one before, but the minimal version didn't have this problem for some reason. I ended up reproducing the issue with this repository: https://github.com/jumoog/intro-skipper/commit/5d743302e41c88f0e8ac85cee88142ca1a235aee

At this specific commit, I can reproduce the issue by clicking "analyze intro" in the scheduled tasks.

**Comment by jumoog** [comment]
https://github.com/jumoog/intro-skipper/blob/5d743302e41c88f0e8ac85cee88142ca1a235aee/ConfusedPolarBear.Plugin.IntroSkipper/Plugin.cs#L218

**Comment by gnattu** [comment]
> I was mainly intrigued about looking into this because this same error was what blocked me when I originally tried to implement plugin unloading. Back then, this was a [bug in the dotnet runtime](https://github.com/dotnet/runtime/issues/1388) (also [this](https://github.com/dotnet/runtime/issues/599) and [this](https://github.com/dotnet/runtime/issues/598) related bug) that caused this error to pop up any time `XmlSerializer` was used in a collectible assembly load context. However, that bug was fixed way back in .NET 6, which allowed someone else to continue on and finish the implementation. So I'm surprised to now see it pop up again.
> 

It shows the same exception but caused by different reasons. The bugs you linked are for all XmlSerializer instances created in the plugin, but this instance is created in the server and passed into the constructor. In your bugs, the plugin cannot even be loaded if it wants to initialize an XmlSerializer. Passing this instance is fine; it won't block the load of the plugin. Using this serializer from the server side is also fine; the server is using it for plugin config files. However, accessing this instance and using its methods to try to serialize/deserialize is not supported because it will cause runtime errors like NullReferenceException or even crashes.

> Perhaps reading through those linked runtime issues might provide some clues on how to get this working. I also came across this documentation article on how to analyze assembly loading issues: https://learn.microsoft.com/en-us/dotnet/core/dependency-loading/collect-details. That all seems pretty complicated/confusing though, so if this quick fix also solves the issue then maybe just go for it 🤷

This should not work by design. The original design of in-process reloading overlooked the problems this approach may cause. This is fine, because even Microsoft overlooked this and have to patch it later in dotnet 7. The way it worked is not supposed to work, and the fact that it actually worked is a bug and should be fixed. We should not try to "get this working" by "analyzing assembly loading issues" because it is not an issue, but a feature limitation. If we really want the plugin to share the serializer with the server, we have to do much more work to replace the XmlSerializer with something that has less runtime magic, such as the DataContractSerializer. However, this obviously requires much more work and will introduce a lot of breaking changes, which is not suitable for 10.9.

**Comment by jumoog** [comment]
In my opinion, this is also important for the 10.9 release.

**Pull request closed without merging**